File: app/src/main/java/org/xdty/callerinfo/application/Application.java
package org.xdty.callerinfo.application;

import android.util.Log;

import com.google.firebase.analytics.FirebaseAnalytics;
import com.tencent.bugly.crashreport.CrashReport;

import org.xdty.callerinfo.di.AppComponent;
import org.xdty.callerinfo.di.DaggerAppComponent;
import org.xdty.callerinfo.di.modules.AppModule;
import org.xdty.callerinfo.model.setting.Setting;
import org.xdty.callerinfo.utils.Alarm;
import org.xdty.callerinfo.utils.Resource;
import org.xdty.callerinfo.utils.Utils;

import javax.inject.Inject;

import cat.ereza.customactivityoncrash.CustomActivityOnCrash;
import io.reactivex.functions.Consumer;
import io.reactivex.plugins.RxJavaPlugins;

public class Application extends android.app.Application {
    public final static String TAG = Application.class.getSimpleName();

    protected AppComponent mAppComponent;

    protected static Application sApplication;

    private FirebaseAnalytics mAnalytics;

    @Inject
    Setting mSetting;

    @Inject
    Alarm mAlarm;

    public AppComponent getAppComponent() {
        return mAppComponent;
    }

    public static Application getApplication() {
        return sApplication;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        sApplication = this;

        RxJavaPlugins.setErrorHandler(new Consumer<Throwable>() {
            @Override
            public void accept(Throwable throwable) throws Exception {
                Log.e(TAG, Log.getStackTraceString(throwable));
            }
        });

        init();
    }

    protected void init() {
        mAppComponent = DaggerAppComponent.builder().appModule(new AppModule(this)).build();

        mAppComponent.inject(this);

        mAnalytics = FirebaseAnalytics.getInstance(this);
        mAnalytics.logEvent(FirebaseAnalytics.Event.APP_OPEN, null);

        Resource.getInstance().init(Utils.changeLang(this));

        if (mSetting.isCatchCrash()) {
            CustomActivityOnCrash.install(this);
        } else {
            CrashReport.initCrashReport(getApplicationContext(), "0eaf845a04", false);
        }

        mSetting.fix();

        mAlarm.alarm();
        mAlarm.enqueueUpgradeWork();
    }
}

File: app/src/main/java/org/xdty/callerinfo/application/Application.kt
package org.xdty.callerinfo.application

import android.annotation.SuppressLint
import android.util.Log
import com.google.firebase.analytics.FirebaseAnalytics
import com.tencent.bugly.crashreport.CrashReport
import io.reactivex.plugins.RxJavaPlugins
import org.xdty.callerinfo.di.AppComponent
import org.xdty.callerinfo.di.DaggerAppComponent
import org.xdty.callerinfo.di.modules.AppModule
import org.xdty.callerinfo.model.setting.Setting
import org.xdty.callerinfo.utils.Alarm
import org.xdty.callerinfo.utils.Resource
import org.xdty.callerinfo.utils.Utils
import javax.inject.Inject

@SuppressLint("Registered")
open class Application : android.app.Application() {
    lateinit var analytics: FirebaseAnalytics
    @Inject
    internal lateinit var setting: Setting
    @Inject
    internal lateinit var alarm: Alarm

    override fun onCreate() {
        super.onCreate()
        application = this
        RxJavaPlugins.setErrorHandler { throwable -> Log.e(TAG, Log.getStackTraceString(throwable)) }
        init()
    }

    protected fun init() {
        appComponent = DaggerAppComponent.builder().appModule(AppModule(this)).build()
        appComponent.inject(this)
        analytics = FirebaseAnalytics.getInstance(this)
        analytics.logEvent(FirebaseAnalytics.Event.APP_OPEN, null)
        Resource.init(Utils.changeLang(this))
        if (!setting.isCatchCrash) {
            CrashReport.initCrashReport(applicationContext, "0eaf845a04", false)
        }
        setting.fix()
        alarm.alarm()
        alarm.enqueueUpgradeWork()
    }

    companion object {
        val TAG = Application::class.java.simpleName
        lateinit var application: Application
        lateinit var appComponent: AppComponent
    }
}